Monorepo RFC (#5)
vtex-apps/rfcs: text/0003-monorepo.md

- Start Date: 2020-07-03
- Implementation PR: (leave this empty)
- Store Discussion Issue: (leave this empty)

# Summary

This RFC proposes a solution to create a single repository to all apps of the Store Framework. 

# Basic example

> We want to do something similar that the following frameworks do:

Reviewer: ```suggestion
We want to do something similar to what the following frameworks do:
```

* [Angular](https://github.com/angular/angular)
* [React](https://github.com/facebook/react)
* [Material UI](https://github.com/mui-org/material-ui)

> They keep all packages in the same repository

Reviewer: ```suggestion
They keep all packages in the same repository.
```

Pain points of keeping 73 repos up to date:

* Lint rules always get out of date in some forgotten apps
* Github templates have the same problem of lint rules
* API patterns and best practices to develop a new block
* Avoid duplication of logic
* Test setup
* Misleading open-source flow to create issues

# Detailed design

Since this is not a feature but an organizational change that might help everyone that works in Store Framework and the open-source community this we don't have much to add to this section, but we expect to fix all the problems listed above and that the monorepo help us to make it easier the following topics:

> * Share code to make testing easier

Reviewer: This point could be expanded. I guess the main reason that testing would be improved is due to the fact that we won't need to mock things anymore, right?

> * Create a calendar of support for majors

Reviewer: We should add examples to this. Maybe the LTS from Node.js?

* Plan new majors for the whole framework
* Create migration guides to each change of these majors
* Predictable manual updates to our clients to get a better store

All these points about majors should have an RFC to explain better how we should do it.

# Drawbacks

Why should we *not* do this? Please consider:

- To launch a new major we would need to launch a major of all apps or to put the legacy major in a separate folder and to add the current in the place of the current major (similar with what the [search-result](https://github.com/vtex-apps/seach-result) does)
- Older major versions that were being kept in different branches would have to make the same strategy to continue to exist
- We would have to update the bot that launches a new version of the apps

> - The repository would be *incredibly* big

Reviewer: I don't know if the repository would be "incredibly big"... there are much bigger ones out there. Maybe even those listed as examples.
I think a looot of the size of our apps has to do with the boilerplate/tooling code. I don't think this repo would be big to the point that we wouldn't be able to clone the entire repo and use it locally, for example.

Author: Is not like that, but people should take an extra care with git now because a bad rebase might create a lot of problems. And the git commands might become a bit slower (no so much)

# Alternatives

If we have a tool that could make the best of two sides, creating a monorepo or using multiple repositories to each app. It would be something easy for the open-source community to adopt and easy for all of us that work on the Store Framework to communicate about API changes and to share code, but this would lead to a lot of work of VTEX too.

# Adoption strategy

We would have to adopt this in phases:

> 1. Prepare the bot to launch new versions in this repository and prepare https://vtex.io/docs to find documentation in the repo

Reviewer: `vtex.io/docs` would have no problem finding the documentation, since it uses VTEX IO's registry 😎

2. Move at least two Store Framework repositories, setup all the tooling considering multiple apps, make sure that everything is working and that this has no implication in production (without sharing code until phase 6)
3. Migrate all apps of the Store Framework team and stay this way for a while as a trial phase

> 4. Migrate one team at a time to the monorepo

Reviewer: Maybe we should expand on this, pointing out the teams that would be involved, so that they already know what we're planning.

Reviewer: We might also have to make exceptions for private repos, such as the ones from `vtex.my-account`

Author: I think we could ask if they still should be private

Reviewer 2: Maybe the monorepo shouldn't have the private repositories. We have to check this out carefully.

Author: It shouldn't, but if we are an open-source framework, all apps that are part of it should be open source too

> 5. Think about how we could share code between apps

Reviewer: This might be a tricky one... since code would code from multiple apps. Maaaybe (big maybe 😂 ) we could even create an app for this shared code... Since all the code would be in the same repo, it doesn't seem like a horrible idea to me

6. Start sharing code
7. Plan next majors how would this work

# How we teach this

We should archive all current repositories and add something in the readme to point to the monorepo.

# Unresolved questions

> How we should make majors and how to make it predictable to the clients when they need to upgrade there store and what would be the benefits of doing it.

Reviewer: To be honest, I think we need an entirely separate RFC for majors.
We could mention it as a motivation for this monorepo one, but I don't think we need to have that figured out in order to create and use the monorepo.

Author: I added this at the end of detailed design, but I think it would be better to move here

- Start Date: 2020-07-03
- Implementation PR: (leave this empty)
- Store Discussion Issue: (leave this empty)

# Summary

This RFC proposes a solution to create a single repository to all apps of the Store Framework. 

# Basic example

We want to do something similar to what the following frameworks do:

* [Angular](https://github.com/angular/angular)
* [React](https://github.com/facebook/react)
* [Material UI](https://github.com/mui-org/material-ui)

They keep all packages in the same repository.

# Motivation

> When the Store Framework started it had only one united team with all the people in the same place and it didn't have many apps (around 12 apps) for this team to support, so it was easy to control who was doing what and if it was being done respecting the standards of the framework, but today we have around 73 repositories that 6-7 different teams of VTEX support and on top of it, now we have an open-source community that uses those apps and even contribute to it. With all this code base it became complicated to update tooling, fix technical depts, and even update these apps to respect best practices of code or API. With a monorepo, we can get closer to the idea of creating a product that has the same ideas across all blocks.

Reviewer 3: Technical Debts

> When the Store Framework started it had only one team united team with all the people in the same place and it had not so many apps (around 12 apps) to this team to support, so it was easy to control who was doing what and if it was being done respecting the standards of the framework, but today we have around 73 repositories that 6-7 different teams of VTEX support and on top of it now we have an open-source community that uses those apps and even contribute to it. With all this code base it became complicated to update tooling, fix technical depts, and even update these apps to respect best practices of code or API. With a monorepo we can get closer to the idea of creating a product that has the same ideas across all blocks.

Reviewer 2: ```suggestion
When the Store Framework started it had only one united team with all the people in the same place and it didn't have many apps (around 12 apps) for this team to support, so it was easy to control who was doing what and if it was being done respecting the standards of the framework, but today we have around 73 repositories that 6-7 different teams of VTEX support and on top of it, now we have an open-source community that uses those apps and even contribute to it. With all this code base it became complicated to update tooling, fix technical depts, and even update these apps to respect best practices of code or API. With a monorepo, we can get closer to the idea of creating a product that has the same ideas across all blocks.
```

Pain points of keeping 73 repos up to date:

* Lint rules always get out of date in some forgotten apps
* Github templates have the same problem of lint rules

> * API patterns and best practices to develop a new block

Reviewer 2: I didn't understand why this is a problem caused by having multiple repositories. Could you explain this in more detail?

Author: This is not the only thing that causes API differences but is one of them, because we don't see all contributions to the framework (and we can't), but with the monorepo, this become a bit easier since all PRs will be in the same place

> * Avoid duplication of logic

Reviewer 2: ```suggestion
* Having duplication of logic
```

* Test setup
* Misleading open-source flow to create issues

# Detailed design

Since this is not a feature but an organizational change that might help everyone that works in Store Framework and the open-source community this we don't have much to add to this section, but we expect to fix all the problems listed above and that the monorepo help us to make it easier the following topics:

> * Share code to make testing easier because we wouldn't need to mock all IO dependencies

Reviewer 2: Amen for Integration Tests 🙌🏼

* Create a calendar of support for majors, like [NodeJS](https://nodejs.org/en/about/releases/)
* Plan new majors for the whole framework
* Create migration guides to each change of these majors

> * Predictable manual updates to our clients to get a better store

Reviewer 2: What do you mean with this?

Author: This is more about the majors for the whole framework topic because if we launch only one major in the whole year our clients only have to pay the agency once every year to change the store, but if we keep launching majors of each app separately it's hard to encourage our clients to update to the newest versions

## What kind of apps should not go to the monorepo

* Pixel apps, e.g. https://github.com/vtex-apps/zendesk-chat
* Template apps, e.g. https://github.com/vtex-apps/react-app-template
* Theme apps, e.g. https://github.com/vtex-apps/store-theme
* Store apps that are out of IO e.g. https://github.com/vtex/checkout-confirmation-ui

# Drawbacks

Why should we *not* do this? Please consider:

- To launch a new major we would need to launch a major of all apps or to put the legacy major in a separate folder and to add the current in the place of the current major (similar with what the [search-result](https://github.com/vtex-apps/seach-result) does)
- Older major versions that were being kept in different branches would have to make the same strategy to continue to exist
- We would have to update the bot that launches a new version of the apps
- The repository would be **incredibly** big

> 

Reviewer 2: Could this affect the store performances too?

Author: No, just the repo that would be really big to download and a bit hard to manage versioning controll

# Alternatives

If we have a tool that could make the best of two sides, creating a monorepo or using multiple repositories to each app. It would be something easy for the open-source community to adopt and easy for all of us that work on the Store Framework to communicate about API changes and to share code, but this would lead to a lot of work of VTEX too.

> 

Reviewer 2: Doesn't `bit` help here? I think they are using it on the gatsby POC

Author: Bit works better with npm packages that's why is being used in the gatsby POC because the apps are npm packages. And I don't know if it has a way to visualize all PRs of an organization or a list of repositories in one single place
